idl/info.go

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package idl
+
+import (
+	"go.uber.org/thriftrw/ast"
+	"go.uber.org/thriftrw/idl/internal"
+)
+
+// Info contains additional information about the parsed document.
+type Info struct {
+	nodePositions internal.NodePositions
+}
+
+// Pos returns a node's position in the parsed document.
+func (i *Info) Pos(n ast.Node) Position {
+	if line := ast.LineNumber(n); line != 0 {
+		return Position{Line: line}
+	}
+	pos := i.nodePositions[n]
+	return Position{Line: pos.Line}
+}
```

Comment:
@abhinav I've just realized this will panic if passed an unhashable node that doesn't support `ast.LineNumber` (like `ast.MapType`). Can you think of a pattern to avoid that here aside from using reflection or `recover()`?

Reply:
Oh, yeah, that's a problem. I think we should not ignore the `!= 0` case here: if the type records its own line number, we should return that as-is. That would narrow down the possibility of panic to only those types that are unhashable, but I don't think we have anything that is unhashable that doesn't record its own line number.

To respect the line number of AST nodes that know their own line number, we have two options: `Info` moves to the AST package (which I suspect is not possible without exposing Info.nodePositions), or we add an `ast.LookupLineNumber(ast.Node) (n int, ok bool)` which reports the line number only if available.

Reply:
Perhaps this is an opportunity to move `Position` to the `ast` package with an accessor that returns `pos, ok`?

Reply:
That sounds reasonable!

Reply:
#504 504